File: web/lib/connection.js
import { supabase } from "./supabase";

export async function newConnection(tid) {
  let currentUser = supabase.auth.user().email;
  let userRes = await supabase
    .from("users")
    .select("uid")
    .match({ email: currentUser });
  const data = {
    fid: userRes.data[0]["uid"],
    tid: tid,
  };
  let res = await supabase.from("connections").insert(data);
  console.log(userRes.data);
  console.log(res.data);
}

export async function getCurrentUser(email) {
  let userRes = await supabase.from("users").select().match({ email: email });
  return userRes.data[0];
}

export async function acceptConnection(tid) {
  let currentUser = supabase.auth.user().email;
  let userRes = await supabase
    .from("users")
    .select("uid")
    .match({ email: currentUser });
  const data = {
    fid: userRes.data[0]["uid"],
    tid: tid,
  };
  let res = await supabase
    .from("connections")
    .update({ connected: true })
    .match(data);
  data = {
    tid: userRes.data[0]["uid"],
    fid: tid,
    connected: true,
  };
  res = await supabase.from("connections").insert(data);
  console.log(userRes.data);
  console.log(res.data);
}

export async function getConnections() {
  let currentUser = supabase.auth.user().email;
  let userRes = await supabase
    .from("users")
    .select("uid")
    .match({ email: currentUser });
  let res = await supabase
    .from("connections")
    .select()
    .match({ fid: userRes.data[0]["uid"] });
  return res.data;
}

export async function pendingConnections() {
  let currentUser = supabase.auth.user().email;
  let userRes = await supabase
    .from("users")
    .select("uid")
    .match({ email: currentUser });
  let res = await supabase
    .from("connections")
    .select()
    .match({ tid: userRes.data[0]["uid"] });
  return res.data;
}

export async function searchUsers(email) {
  let res = await supabase.from("users").select("uid").match({ email: email });
  if (res.data.length == 0) {
    return -1;
  } else {
    return res.data[0]["uid"];
  }
}

export async function rejectConnection(tid) {
  let currentUser = supabase.auth.user().email;
  let userRes = await supabase
    .from("users")
    .select("uid")
    .match({ email: currentUser });
  const data = {
    fid: userRes.data[0]["uid"],
    tid: tid,
  };

  let res = await supabase.from("connections").delete().match(data);
  console.log(userRes.data);
  console.log(res.data);
}

export async function getUser(tid) {
  let res = await supabase.from("users").select().match({ uid: tid });
  return res.data[0];
}

export async function signOut() {
  let res = await supabase.auth.signOut();
  console.log(res.data);
}

File: lib/services/supabase/connection.dart
import 'constants.dart';

Future<bool> isConnected(int tid) async {
  final res = await supabase
      .from('connections')
      .select('connected')
      .match({'tid': tid}).execute();
  List<dynamic> data = res.data;
  return data[0]['connected'];
}

Future<void> newConnection(int tid) async {
  var currentUser = await supabase.auth.currentUser!.email;
  var userRes = await supabase
      .from('users')
      .select('uid')
      .match({'email': currentUser}).execute();
  final data = {
    'fid': userRes.data[0]['uid'],
    'tid': tid,
  };
  final res = await supabase.from('connections').insert(data).execute();
  print(userRes.data);
  print(res.data);
}

Future<void> rejectConnection(int fid) async {
  var currentUser = await supabase.auth.currentUser!.email;
  var userRes = await supabase
      .from('users')
      .select('uid')
      .match({'email': currentUser}).execute();
  final data = {
    'fid': fid,
    'tid': userRes.data[0]['uid'],
  };
  // creates a pending request
  final res = await supabase.from('connections').delete().match(data).execute();
  print(userRes.data);
  print(res.data);
}

Future<void> acceptConnection(int fid) async {
  var currentUser = await supabase.auth.currentUser!.email;
  var userRes = await supabase
      .from('users')
      .select('uid')
      .match({'email': currentUser}).execute();
  var data = {
    'fid': fid,
    'tid': userRes.data[0]['uid'],
  };

  // updates the connected status to true
  var res = await supabase
      .from('connections')
      .update({'connected': true})
      .match(data)
      .execute();
  data = {
    'tid': fid,
    'fid': userRes.data[0]['uid'],
    'connected': true,
  };
  // creates a two way connection between the two users
  res = await supabase.from('connections').insert(data).execute();
  print(userRes.data);
  print(res.data);
}

Future<int?> searchConnection(String email) async {
  var res = await supabase
      .from('users')
      .select('uid')
      .match({'email': email}).execute();
  print(res.data);
  // returns the user id if found
  return (res.data as List<dynamic>).isNotEmpty ? res.data[0]['uid'] : null;
}

Future<List<dynamic>> getConnections() async {
  var currentUser = await supabase.auth.currentUser!.email;
  var userRes = await supabase
      .from('users')
      .select('uid')
      .match({'email': currentUser}).execute();
  var data = {
    'fid': userRes.data[0]['uid'],
    'connected': true,
  };
  var res = await supabase.from('connections').select().match(data).execute();
  return res.data;
}

Future<List<dynamic>> pendingConnections() async {
  var currentUser = await supabase.auth.currentUser!.email;
  var userRes = await supabase
      .from('users')
      .select('uid')
      .match({'email': currentUser}).execute();
  var data = {
    'tid': userRes.data[0]['uid'],
    'connected': false,
  };
  var res = await supabase.from('connections').select().match(data).execute();
  return res.data;
}
